# Problem 2

In [19]:
# Import Libraries
import pandas as pd
import itertools

In [24]:
# Load dataset
df = pd.read_csv("synthetic_wtp_laptop_data.csv")

# Define base model and upgrade costs
base_model = {
    'Memory': 16,
    'Storage': 512,
    'CPU_class': 1,
    'Screen_size': 14.0,
    'year': 2025,
    'price': 111000
}

upgrade_options = {
    'Add 16 GB memory': {'Memory': 32, 'Cost': 7000},
    'Add 512 GB storage': {'Storage': 1024, 'Cost': 5000},
    'Upgrade CPU class by 1': {'CPU_class': 2, 'Cost': 15000},
    'Increase screen size to 16': {'Screen_size': 16.0, 'Cost': 3000}
}

# Function to apply upgrades to base model
def apply_upgrades(base, upgrades):
    updated = base.copy()
    total_cost = 0
    for name in upgrades:
        for key, val in upgrade_options[name].items():
            updated[key] = val
        total_cost += upgrade_options[name]['Cost']
    return updated, total_cost

# Check all 2-combination upgrades
results = []
for combo in combinations(upgrade_options.keys(), 2):
    # Apply upgrades to base model
    upgraded_model, cost = apply_upgrades(base_model, combo)
    
    # Filter dataset to match upgraded model specs
    filtered = df[
        (df['Memory'] == upgraded_model['Memory']) &
        (df['Storage'] == upgraded_model['Storage']) &
        (df['CPU_class'] == upgraded_model['CPU_class']) &
        (df['Screen_size'] == upgraded_model['Screen_size']) &
        (df['year'] == upgraded_model['year'])
    ]
    
    # If no matching rows, skip
    if len(filtered) == 0:
        continue
    
    avg_price = filtered['price'].mean()
    gross_profit = avg_price - base_model['price'] - cost
    
    results.append({
        'Upgrades': combo,
        'Avg_Selling_Price': round(avg_price),
        'Upgrade_Cost': cost,
        'Gross_Profit': round(gross_profit)
    })

# Convert to DataFrame and sort
results_df = pd.DataFrame(results)
results_df = results_df.sort_values(by='Gross_Profit', ascending=False)

# Display the results
results_df.reset_index(drop=True, inplace=True)
results_df


Upgrades  Avg_Selling_Price  \
0         (Add 16 GB memory, Upgrade CPU class by 1)             158504   
1       (Add 512 GB storage, Upgrade CPU class by 1)             146755   
2  (Upgrade CPU class by 1, Increase screen size ...             144164   
3             (Add 16 GB memory, Add 512 GB storage)             137498   
4     (Add 16 GB memory, Increase screen size to 16)             133770   
5   (Add 512 GB storage, Increase screen size to 16)             121602   

   Upgrade_Cost  Gross_Profit  
0         22000         25504  
1         20000         15755  
2         18000         15164  
3         12000         14498  
4         10000         12770  
5          8000          2602

# Solution

Add 16 GB memory, Upgrade CPU class by 1

Based on the analysis of the dataset, the combination of **adding 16 GB of memory** and **upgrading the CPU class by 1 level** yields the highest gross profit among all possible two-upgrade options. This configuration results in an estimated average selling price of ¥158,504, with a total upgrade cost of ¥22,000, leading to a gross profit of **¥25,504**. This indicates that customers value enhanced performance features such as increased memory and a more powerful CPU the most, making this the optimal choice for maximizing profitability in the new product lineup.